src/Microsoft.AspNetCore.Sockets.Client/Connection.cs

```diff
@@ -14,70 +14,174 @@ namespace Microsoft.AspNetCore.Sockets.Client
+                await StartTransport(connectUrl);
+            }
+            catch
+            {
+                Interlocked.Exchange(ref _connectionState, ConnectionState.Disconnected);
+                throw;
+            }
+
+            // start receive loop
+            _receiveLoopTask = ReceiveAsync();
+
+            Interlocked.Exchange(ref _connectionState, ConnectionState.Connected);
+
+            // Do not "simplify" - events can be removed from a different thread
+            var connectedEventHandler = Connected;
+            if (connectedEventHandler != null)
+            {
+                connectedEventHandler();
+            }
         }
 
-        public async Task<bool> ReceiveAsync(ReceiveData receiveData, CancellationToken cancellationToken)
+        private static async Task<Uri> GetConnectUrl(Uri url, HttpClient httpClient, ILogger logger)
         {
-            if (receiveData == null)
+            var disposeHttpClient = httpClient == null;
+            httpClient = httpClient ?? new HttpClient();
+            try
             {
-                throw new ArgumentNullException(nameof(receiveData));
+                var connectionId = await GetConnectionId(url, httpClient, logger);
+                return Utils.AppendQueryString(url, "id=" + connectionId);
             }
+            finally
+            {
+                if (disposeHttpClient)
+                {
+                    httpClient.Dispose();
+                }
+            }
+        }
 
-            if (Input.Completion.IsCompleted)
+        private static async Task<string> GetConnectionId(Uri url, HttpClient httpClient, ILogger logger)
+        {
+            var negotiateUrl = Utils.AppendPath(url, "negotiate");
+            try
+            {
+                // Get a connection ID from the server
+                logger.LogDebug("Establishing Connection at: {0}", negotiateUrl);
+                var connectionId = await httpClient.GetStringAsync(negotiateUrl);
+                logger.LogDebug("Connection Id: {0}", connectionId);
+                return connectionId;
+            }
+            catch (Exception ex)
+            {
+                logger.LogError("Failed to start connection. Error getting connection id from '{0}': {1}", negotiateUrl, ex);
+                throw;
+            }
+        }
+
+        private async Task StartTransport(Uri connectUrl)
+        {
+            var applicationToTransport = Channel.CreateUnbounded<Message>();
+            var transportToApplication = Channel.CreateUnbounded<Message>();
+            var applicationSide = new ChannelConnection<Message>(transportToApplication, applicationToTransport);
+
+            _transportChannel = new ChannelConnection<Message>(applicationToTransport, transportToApplication);
+#pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
```

Comment:
You can even do `_ = [expression]` in C# now, you don't need to declare a variable.